[Issue] Linux ASAN CI: storage-libs / hipfile build fails — "undefined symbols pthread_getspecific/pthread_key_create/pthread_setspecific/pthread_key_delete (missing Threads::Threads linkage in hipfile test CMakeLists)"

## Context

- **Workflow:** Multi-Arch CI ASAN (`.github/workflows/multi_arch_ci_asan.yml`)
- **Run:** [Run 27662811532 (2026-06-17)](https://github.com/ROCm/TheRock/actions/runs/27662811532)
- **Failing job:** [Linux::asan / Build Multi-Arch Stages / storage-libs / Stage - Storage Libs](https://github.com/ROCm/TheRock/actions/runs/27662811532/job/81812830420)
- **Affected arch:** N/A (host-side link failure, no GPU required)
- **First failing run:** [Run 27662811532 (2026-06-17)](https://github.com/ROCm/TheRock/actions/runs/27662811532) — first appearance of the storage-libs job, introduced by [PR #5217](https://github.com/ROCm/TheRock/pull/5217)
- **Last passing run:** N/A — job never previously existed
- **Regression commit:** [`bbd306dc169e`](https://github.com/ROCm/TheRock/commit/bbd306dc169ea699b714f04e36e1ce4a24876a23) — "[hipfile] Add initial hipFile support"

## Summary

The `storage-libs / hipfile` stage fails during the link phase for all three test executables (`hipfile_tests`, `hipfile_system_tests`, `test/amd_detail/internal_tests`) with undefined symbol errors from `libgtest.a` / `libgmock.a`:

```text
[hipfile] FAILED: test/hipfile_tests
[hipfile] ld.lld: error: undefined symbol: pthread_getspecific
[hipfile] >>> referenced by gtest-port.h:1774
[hipfile] >>>               gtest-all.cc.o:(testing::internal::UnitTestImpl::~UnitTestImpl()) in archive .../libgtest.a
[hipfile] ld.lld: error: undefined symbol: pthread_key_delete
[hipfile] ld.lld: error: undefined symbol: pthread_key_create
[hipfile] ld.lld: error: undefined symbol: pthread_setspecific
[hipfile] clang++: error: linker command failed with exit code 1
```

## Root Cause

The hipfile test CMakeLists files in `rocm-systems` (pinned at `7ced1a6`) do not link `Threads::Threads` (i.e., `-lpthread`) for the test executables:

- `projects/hipfile/test/CMakeLists.txt` — `hipfile_tests` and `hipfile_system_tests` link only against `GTest::gtest` (and `GTest::gtest_main`), with no `Threads::Threads`
- `projects/hipfile/test/amd_detail/CMakeLists.txt` — calls `find_package(Threads REQUIRED)` but does not add `target_link_libraries(internal_tests PRIVATE Threads::Threads)`

`ld.lld` (LLVM's linker, used in the ASAN toolchain) is strict about undefined symbols in static archives. GTest's `ThreadLocal<T>` implementation uses POSIX thread-key APIs (`pthread_key_create`, etc.) from `gtest-port.h:1774–1819`, which require explicit `-lpthread` linkage under lld. GNU ld resolves these transitively from glibc, masking the bug in non-ASAN builds.

@derobins  @gaoikawa 
Still noticing these failures in rocm-systems bump(https://github.com/ROCm/TheRock/pull/5946) which is carrying the fix rocm-system#7387 
commit diff: https://github.com/ROCm/rocm-systems/compare/7ced1a660f03b3cf06197238692485f73c4772b9...c59673fc3268156835d9f0f48215462ad53c31d5

log: https://github.com/ROCm/TheRock/actions/runs/27729666076/job/82053632711?pr=5946

```
__w/TheRock/TheRock/build/core/clr/dist/lib/llvm/lib/clang/23/lib/linux/libclang_rt.builtins-x86_64.a && :
[hipfile] ld.lld: error: undefined symbol: pthread_getspecific
[hipfile] >>> referenced by gtest-port.h:1774 (/__w/TheRock/TheRock/build/third-party/googletest/source/googletest/include/gtest/internal/gtest-port.h:1774)
[hipfile] >>>               gtest-all.cc.o:(testing::internal::UnitTestImpl::~UnitTestImpl()) in archive /__w/TheRock/TheRock/build/third-party/googletest/dist/lib/libgtest.a
[hipfile] >>> referenced by gtest-port.h:1774 (/__w/TheRock/TheRock/build/third-party/googletest/source/googletest/include/gtest/internal/gtest-port.h:1774)
[hipfile] >>>               gtest-all.cc.o:(testing::internal::UnitTestImpl::~UnitTestImpl()) in archive /__w/TheRock/TheRock/build/third-party/googletest/dist/lib/libgtest.a
[hipfile] >>> referenced by gtest-port.h:1812 (/__w/TheRock/TheRock/build/third-party/googletest/source/googletest/include/gtest/internal/gtest-port.h:1812)
[hipfile] >>>               gmock-all.cc.o:(testing::InSequence::InSequence()) in archive /__w/TheRock/TheRock/build/third-party/googletest/dist/lib/libgmock.a
[hipfile] >>> referenced 10 more times
[hipfile] 
[hipfile] ld.lld: error: undefined symbol: pthread_key_delete
[hipfile] >>> referenced by gtest-port.h:1778 (/__w/TheRock/TheRock/build/third-party/googletest/source/googletest/include/gtest/internal/gtest-port.h:1778)
[hipfile] >>>               gtest-all.cc.o:(testing::internal::UnitTestImpl::~UnitTestImpl()) in archive /__w/TheRock/TheRock/build/third-party/googletest/dist/lib/libgtest.a
[hipfile] >>> referenced by gtest-port.h:1778 (/__w/TheRock/TheRock/build/third-party/googletest/source/googletest/include/gtest/internal/gtest-port.h:1778)
[hipfile] >>>               gtest-all.cc.o:(testing::internal::UnitTestImpl::~UnitTestImpl()) in archive /__w/TheRock/TheRock/build/third-party/googletest/dist/lib/libgtest.a
[hipfile] >>> referenced by gtest-port.h:1778 (/__w/TheRock/TheRock/build/third-party/googletest/source/googletest/include/gtest/internal/gtest-port.h:1778)
[hipfile] >>>               gmock-all.cc.o:(testing::internal::ThreadLocal<testing::Sequence*>::~ThreadLocal()) in archive /__w/TheRock/TheRock/build/third-party/googletest/dist/lib/libgmock.a
[hipfile] >>> referenced 2 more times
[hipfile] 
[hipfile] ld.lld: error: undefined symbol: pthread_setspecific
[hipfile] >>> referenced by gtest-port.h:1819 (/__w/TheRock/TheRock/build/third-party/googletest/source/googletest/include/gtest/internal/gtest-port.h:1819)
[hipfile] >>>               gmock-all.cc.o:(testing::InSequence::InSequence()) in archive /__w/TheRock/TheRock/build/third-party/googletest/dist/lib/libgmock.a
[hipfile] >>> referenced by gtest-port.h:1819 (/__w/TheRock/TheRock/build/third-party/googletest/source/googletest/include/gtest/internal/gtest-port.h:1819)
[hipfile] >>>               gmock-all.cc.o:(testing::InSequence::InSequence()) in archive /__w/TheRock/TheRock/build/third-party/googletest/dist/lib/libgmock.a
[hipfile] >>> referenced by gtest-port.h:1819 (/__w/TheRock/TheRock/build/third-party/googletest/source/googletest/include/gtest/internal/gtest-port.h:1819)
[hipfile] >>>               gmock-all.cc.o:(testing::InSequence::~InSequence()) in archive /__w/TheRock/TheRock/build/third-party/googletest/dist/lib/libgmock.a
[hipfile] >>> referenced 5 more times
[hipfile] 
[hipfile] ld.lld: error: undefined symbol: pthread_key_create
[hipfile] >>> referenced by gtest-port.h:1805 (/__w/TheRock/TheRock/build/third-party/googletest/source/googletest/include/gtest/internal/gtest-port.h:1805)
[hipfile] >>>               gtest-all.cc.o:(testing::internal::UnitTestImpl::UnitTestImpl(testing::UnitTest*)) in archive /__w/TheRock/TheRock/build/third-party/googletest/dist/lib/libgtest.a
[hipfile] >>> referenced by gtest-port.h:1805 (/__w/TheRock/TheRock/build/third-party/googletest/source/googletest/include/gtest/internal/gtest-port.h:1805)
[hipfile] >>>               gtest-all.cc.o:(testing::internal::UnitTestImpl::UnitTestImpl(testing::UnitTest*)) in archive /__w/TheRock/TheRock/build/third-party/googletest/dist/lib/libgtest.a
[hipfile] >>> referenced by gtest-port.h:1805 (/__w/TheRock/TheRock/build/third-party/googletest/source/googletest/include/gtest/internal/gtest-port.h:1805)
[hipfile] >>>               gmock-all.cc.o:(_GLOBAL__sub_I_gmock_all.cc) in archive /__w/TheRock/TheRock/build/third-party/googletest/dist/lib/libgmock.a
[hipfile] clang++: error: linker command failed with exit code 1 (use -v to see invocation)
[hipfile] ninja: build stopped: subcommand failed.
ninja: build stopped: cannot make progress due to previous errors.
Error: Process completed with exit code 1.
0s
0s
0s
```

--- a/projects/hipfile/test/CMakeLists.txt
+++ b/projects/hipfile/test/CMakeLists.txt
@@ -47,6 +47,8 @@ ais_add_test_executable(
 )
 target_link_libraries(hipfile_tests PRIVATE GTest::gtest)
 target_link_libraries(hipfile_tests PRIVATE GTest::gtest_main)
+target_compile_options(hipfile_tests PRIVATE -pthread)
+target_link_options(hipfile_tests PRIVATE -pthread)
 
 ais_gtest_discover_tests(
     hipfile_tests
@@ -61,6 +63,8 @@ ais_add_test_executable(
     SYSINCLS ${TEST_SYSINCLS}
 )
 target_link_libraries(hipfile_system_tests PRIVATE GTest::gtest)
+target_compile_options(hipfile_system_tests PRIVATE -pthread)
+target_link_options(hipfile_system_tests PRIVATE -pthread)
 
 ais_gtest_discover_tests(
     hipfile_system_tests

--- a/projects/hipfile/test/amd_detail/CMakeLists.txt
+++ b/projects/hipfile/test/amd_detail/CMakeLists.txt
@@ -50,6 +50,8 @@ ais_add_test_executable(
 # Add gtest
 target_link_libraries(internal_tests PRIVATE GTest::gmock)
 target_link_libraries(internal_tests PRIVATE GTest::gtest)
+target_compile_options(internal_tests PRIVATE -pthread)
+target_link_options(internal_tests PRIVATE -pthread)
 
 ais_gtest_discover_tests(
     internal_tests

